File: 2022/Day4/main.exs
defmodule Day4 do
  @moduledoc """
  Code for completing Day 4 of Advent of Code 2022
  """

  @doc """
  Returns converts a single line to four integers representing two ranges
  """
  def line_to_ranges(line) do
    splits = String.split(line, ",")
    ranges_str = String.split(Enum.at(splits, 0), "-") ++ String.split(Enum.at(splits, 1), "-") #as strings
    ranges_str |> Enum.map(fn x -> Integer.parse(x) |> elem(0) end) #as integers
  end


  @doc """
  Returns a bool saying if one of the two ranges fully contains the other
  """
  def fully_contains?(ranges) do
    [elem0, elem1, elem2, elem3 | _] = ranges #extract the four into individual variables

    #bot0     bot1     top0     top1       bot1     bot0     top1     top0
    (elem0 <= elem2 && elem1 >= elem3) || (elem2 <= elem0 && elem3 >= elem1)

  end


  @doc """
  Returns a bool saying if the two ranges overlap
  """
  def overlap?(ranges) do
    [elem0, elem1, elem2, elem3 | _] = ranges #extract the four into individual variables

    #bot0     bot1     top0     bot1       bot1     bot0     top1     bot0
    (elem0 <= elem2 && elem1 >= elem2) || (elem2 <= elem0 && elem3 >= elem0)
  end


  @doc """
  Main entry point
  """
  def main do

    # first retrieve the text from the file
    {result, file} = File.read("input.txt")

    # filter out any errors
    if result == :error do
      IO.puts "Ran into an error opening the input file"
      :error
    else

      # split the file into lines
      ranges = String.split(file, "\n") |> Enum.map(&line_to_ranges/1)

      # filter all the lines
      contain = ranges |> Enum.filter(&fully_contains?/1)
      overlap = ranges |> Enum.filter(&overlap?/1)

      # print the outputs
      IO.puts length(contain)
      IO.puts length(overlap)

      :ok
    end
  end

end

Day4.main

File: 2022/Day4/main.hs
import System.IO
import qualified Data.List as List
import Data.Maybe


-- | Splits a line into four integers
splitToTuple :: [Char] -> (Int, Int, Int, Int)
splitToTuple line = do
    let ind = fromJust $ List.elemIndex ',' line
    let (x,comma:y) = List.splitAt ind line
    let indx = fromJust $ List.elemIndex '-' x 
    let (x0,dash0:y0) = List.splitAt indx x
    let indy = fromJust $ List.elemIndex '-' y 
    let (x1,dash1:y1) = List.splitAt indy y 
    (read x0 :: Int, read y0 :: Int, read x1 :: Int, read y1 :: Int)


-- | Returns whether one set of coordinates fully contains the other
isContained :: (Int, Int, Int, Int) -> Bool
isContained (x, y, z, a) = (x <= z && y >= a) || (z <= x && a >= y)


-- | Returns whether the coordinates overlap at all
isOverlapping :: (Int, Int, Int, Int) -> Bool
isOverlapping (x, y, z, a) = (x <= a && y >= z) || (z <= y && a >= x)


-- | Main entry point
main :: IO ()
main = do
    handle <- openFile "input.txt" ReadMode
    contents <- hGetContents handle
    let all_lines = lines contents
    let ints = map splitToTuple all_lines
    print $ length $ filter isContained ints
    print $ length $ filter isOverlapping ints

